packages/cozy-stack-client/src/OAuthClientsCollection.js

```diff
@@ -0,0 +1,122 @@
+import get from 'lodash/get'
+
+import DocumentCollection from './DocumentCollection'
+import { uri } from './utils'
+import * as querystring from './querystring'
+import { dontThrowNotFoundError } from './Collection'
+import { FetchError } from './errors'
+
+export const OAUTH_CLIENTS_DOCTYPE = 'io.cozy.oauth.clients'
+
+const normalizeDoc = DocumentCollection.normalizeDoctypeJsonApi(
+  OAUTH_CLIENTS_DOCTYPE
+)
+const normalizeOAuthClient = client => ({
+  ...normalizeDoc(client, OAUTH_CLIENTS_DOCTYPE),
+  ...client.attributes
+})
+
+/**
+ * Implements `DocumentCollection` API to interact with the /settings/clients endpoint of the stack
+ */
+class OAuthClientsCollection extends DocumentCollection {
+  constructor(stackClient) {
+    super(OAUTH_CLIENTS_DOCTYPE, stackClient)
+  }
+
+  /**
+   * Fetches all OAuth clients
+   *
+   * @param  {object} options           Query options
+   * @param  {number} options.limit     For pagination, the number of results to return.
+   * @param  {object} options.bookmark  For cursor-based pagination, the index cursor.
+   * @returns {object}                  The JSON API conformant response.
+   */
+  async all(options = {}) {
+    const { limit = 100, bookmark, keys } = options
+
+    const params = {
+      'page[limit]': limit,
+      'page[cursor]': bookmark
+    }
+    const url = uri`/settings/clients`
+    const path = querystring.buildURL(url, params)
+    let resp
+    try {
+      resp = await this.stackClient.fetchJSON('GET', path)
+    } catch (error) {
+      return dontThrowNotFoundError(error)
+    }
+    const nextLink = get(resp, 'links.next', '')
+    const nextLinkURL = new URL(`${this.stackClient.uri}${nextLink}`)
+    const nextBookmark = nextLinkURL.searchParams.get('page[cursor]')
+
+    if (keys) {
+      const data = resp.data
+        .filter(c => keys.includes(c.id))
+        .map(c => normalizeOAuthClient(c))
+      const meta = { ...resp.meta, count: data.length }
+
+      return {
+        data,
+        meta,
+        next: keys.length > data.length && nextBookmark != null,
+        bookmark: nextBookmark || undefined
+      }
+    } else {
+      return {
+        data: resp.data.map(c => normalizeOAuthClient(c)),
+        meta: resp.meta,
+        next: nextBookmark != null,
+        bookmark: nextBookmark || undefined
```

Comment:
nit: as this code is repeated I would have introduced a `hasBookmark` var just after `nextBookmark` declaration. And also i would have converted `nextBookmark` to undefined directly on its initialization. This would avoid usage of `null` and `undefined` that differs depending on the context.